runtime/frame_descriptors.c

```diff
@@ -168,64 +189,204 @@ void caml_init_frame_descriptors(void)
   for (int i = 0; caml_frametable[i] != 0; i++)
     frametables = cons(caml_frametable[i], frametables);
 
+  CAMLassert(frametables != NULL);
+
   /* `init_frame_descriptors` is called from `init_gc`, before
      any mutator can run. We can mutate [current_frame_descrs]
      at will. */
-  add_frame_descriptors(&current_frame_descrs, frametables);
+  realloc_frame_descriptors(&current_frame_descrs, frametables);
+  caml_plat_mutex_init(&frame_mutex);
 }
 
+typedef struct {
+  caml_frame_descrs *table;
+  caml_frametable_list *new_frametables;
+  caml_frametable_list *tail;
+  intnat increase;
+} realloc_request;
 
-typedef struct frametable_array {
-  void **table;
-  int ntables;
-} frametable_array;
+static void realloc_frame_descriptors_from_stw_single(
+    caml_domain_state* domain,
+    void* data,
+    int participating_count,
+    caml_domain_state** participating)
+{
+  barrier_status b = caml_global_barrier_begin ();
+
+  if (caml_global_barrier_is_final(b)) {
+    realloc_request *request = data;
+    caml_frame_descrs *table = request->table;
+    caml_frametable_list *new_frametables = request->new_frametables;
+    caml_frametable_list *tail = request->tail;
+    intnat tblsize = capacity(*table), increase = request->increase;
+
+    if(tblsize < (table->num_descr + increase) * 2) {
+      /* Merge both lists */
+      tail->next = table->frametables;
+
+      realloc_frame_descriptors(table, new_frametables);
+    } else {
+      table->num_descr += increase;
+      fill_hashtable(table, new_frametables);
+      tail->next = table->frametables;
+      table->frametables = new_frametables;
+    }
+  }
+
+  caml_global_barrier_end(b);
+}
 
-static void register_frametables_from_stw_single(frametable_array *array)
+static void add_frame_descriptors(
+  caml_frame_descrs *table,
+  caml_frametable_list *new_frametables)
+{
+  CAMLassert(new_frametables != NULL);
+
+  caml_frametable_list *tail = frametables_list_tail(new_frametables);
+  intnat increase = count_descriptors(new_frametables);
+  intnat tblsize = capacity(*table);
+
+  caml_plat_lock(&frame_mutex);
```

Comment:
A minor point: this function is written to take a `caml_frame_descrs *` state, instead of acting on the `current_frame_descrs` global. Why have a single global mutex? I would expect the mutex to be stored as part of the `caml_caml_descrs` structure. (Or we give up on the `table` parameter and make the function act on `current_frame_descrs` directly; but I like the idea of packing the global state in a structure to easily delineate it.)

Reply:
Good point. I like the idea to put the mutex inside the `caml_frame_descrs` structure.